src/lib/libutil/opendev.3
.\"	$OpenBSD: opendev.3,v 1.21 2013/06/05 03:40:26 tedu Exp $
.\"
.\" Copyright (c) 2000, Todd C. Miller.  All rights reserved.
.\" Copyright (c) 1996, Jason Downs.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS
.\" OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
.\" CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd $Mdocdate: June 5 2013 $
.Dt OPENDEV 3
.Os
.Sh NAME
.Nm opendev
.Nd short form device open routine
.Sh SYNOPSIS
.In util.h
.Ft int
.Fn opendev "const char *path" "int oflags" "int dflags" "char **realpath"
.Sh DESCRIPTION
The
.Fn opendev
function opens a device using a
.Dq short form
name or
.Xr disklabel 8
UID
.Pq DUID .
For instance
.Dq sd0
or
.Dq sd0c
will be expanded to
.Pa /dev/rsd0c
on most architectures.
.Pp
Device name lookup is done by first checking
.Fa path
for a
.Sq /
and if one is found attempting to open that file.
If not,
.Fa path
is checked to see if it is a valid DUID.
If it is, the corresponding device is obtained via
.Xr diskmap 4 .
If
.Fa path
has no
.Sq /
and is not a DUID,
.Fa /dev
is searched for a matching device.
.Pp
The
.Fa oflags
are the same as the
.Fa flags
passed to
.Xr open 2 .
.Pp
The
.Fa dflags
are specified by OR'ing the following values:
.Bd -literal -offset indent
OPENDEV_PART	attempt to open the raw partition during expansion
OPENDEV_BLCK	open the block device (default is character device)
.Ed
.Pp
If
.Fa realpath
is not
.Dv NULL ,
it is modified to point at the fully expanded device name.
.Sh RETURN VALUES
The
.Fn opendev
return value and errors are the same as the return value and errors of
.Xr open 2 .
.Sh SEE ALSO
.Xr open 2 ,
.Xr getrawpartition 3 ,
.Xr diskmap 4 ,
.Xr disklabel 8
.Sh HISTORY
The
.Fn opendev
function first appeared in
.Ox 1.2 .
.Sh CAVEATS
If
.Fa realpath
is not
.Dv NULL ,
on return it will point to internal
static storage space that will be overwritten by subsequent calls.


src/lib/libutil/opendev.c
/*	$OpenBSD: opendev.c,v 1.13 2010/12/21 15:47:52 millert Exp $	*/

/*
 * Copyright (c) 2000, Todd C. Miller.  All rights reserved.
 * Copyright (c) 1996, Jason Downs.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS
 * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <paths.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include <sys/ioctl.h>
#include <sys/limits.h>
#include <sys/disk.h>
#include <sys/dkio.h>

#include "util.h"

/*
 * This routine is a generic rewrite of the original code found in
 * disklabel(8).
 */
int
opendev(const char *path, int oflags, int dflags, char **realpath)
{
	static char namebuf[PATH_MAX];
	struct dk_diskmap dm;
	char *slash, *prefix;
	int fd;

	/* Initial state */
	fd = -1;
	errno = ENOENT;

	if (dflags & OPENDEV_BLCK)
		prefix = "";			/* block device */
	else
		prefix = "r";			/* character device */

	if ((slash = strchr(path, '/'))) {
		strlcpy(namebuf, path, sizeof(namebuf));
		fd = open(namebuf, oflags);
	} else if (isduid(path, dflags)) {
		strlcpy(namebuf, path, sizeof(namebuf));
		if ((fd = open("/dev/diskmap", oflags)) != -1) {
			bzero(&dm, sizeof(struct dk_diskmap));
			dm.device = namebuf;
			dm.fd = fd;
			if (dflags & OPENDEV_PART)
				dm.flags |= DM_OPENPART;
			if (dflags & OPENDEV_BLCK)
				dm.flags |= DM_OPENBLCK;

			if (ioctl(fd, DIOCMAP, &dm) == -1) {
				close(fd);
				fd = -1;
				errno = ENOENT;
			}
		}
	}
	if (!slash && fd == -1 && errno == ENOENT) {
		if (dflags & OPENDEV_PART) {
			/*
			 * First try raw partition (for removable drives)
			 */
			if (snprintf(namebuf, sizeof(namebuf), "%s%s%s%c",
			    _PATH_DEV, prefix, path, 'a' + getrawpartition())
			    < sizeof(namebuf)) {
				fd = open(namebuf, oflags);
			} else
				errno = ENAMETOOLONG;
		}
		if (fd == -1 && errno == ENOENT) {
			if (snprintf(namebuf, sizeof(namebuf), "%s%s%s",
			    _PATH_DEV, prefix, path) < sizeof(namebuf)) {
				fd = open(namebuf, oflags);
			} else
				errno = ENAMETOOLONG;
		}
	}
	if (realpath)
		*realpath = namebuf;

	return (fd);
}
